chore(flags): Add canonical log line for /flags requests

haacked opened the pull request:
## Problem

When debugging `/flags` request issues in production, we need to correlate multiple log entries to understand what happened during a request. This makes debugging slow and error-prone, especially when investigating SDK-specific issues or tracking down problems for specific teams.

We need a single, comprehensive log entry per request that contains all key telemetry. This is inspired by this post entitled [Fast and flexible observability with canonical log lines](https://stripe.com/blog/canonical-log-lines). @andyzzhao brought this post to our attention in [this Slack thread](https://posthog.slack.com/archives/C07Q2U4BH4L/p1766351262135789) and noted that it would enable simple ClickHouse queries for debugging without needing to join multiple log entries.

## Changes

Implement Stripe-style canonical log lines for the Rust `/flags` service:

**Core canonical log infrastructure:**

- Add `CanonicalLogLine` struct to accumulate telemetry throughout request lifecycle
- Add `CanonicalLogGuard` RAII wrapper to ensure logs are always emitted, even on early returns
- Emit a single log entry at request completion with message `"canonical_log_line"`

**Fields captured:**
| Field | Description |
|-------|-------------|
| `request_id` | UUID for distributed tracing |
| `team_id` | Team identifier |
| `distinct_id` | User identifier (truncated to 64 chars) |
| `ip` | Client IP address |
| `user_agent` | Full User-Agent string |
| `lib` | SDK name (e.g., "posthog-python", "web") |
| `lib_version` | SDK version (from query param or User-Agent) |
| `api_version` | Response format version (v=1,2,3,4) |
| `duration_ms` | Request latency |
| `http_status` | Response status code |
| `flags_evaluated` | Count of flags evaluated |
| `flags_enabled` | Count of enabled flags |
| `flags_disabled` | Whether flags were disabled via request |
| `quota_limited` | Hit quota limit |
| `rate_limited` | Hit rate limit |
| `error_code` | Error code if request failed |

**SDK detection:**
- Browser SDK: `lib_version` from `ver=` query param, `lib` = "web" from User-Agent patterns
- Server SDKs: Both `lib` and `lib_version` extracted from User-Agent (e.g., `posthog-python/3.0.0`)

## How did you test this code?

- Unit tests for all helper functions (`extract_version_from_user_agent`, `extract_lib_from_user_agent`, `extract_client_ip`, `extract_request_id`)
- Unit tests for `CanonicalLogLine` struct initialization and field population
- Unit tests for `CanonicalLogGuard` RAII behavior (emit on drop, explicit emit, into_inner)
- Manual testing with curl to verify log output format

To filter canonical logs when running locally:
```bash
RUST_LOG=info cargo run --bin feature-flags 2>&1 | grep "canonical_log_line"
```

Example output:

```
2025-12-24T00:20:20.775160Z  INFO ThreadId(09) feature_flags::handler::canonical_log: canonical_log_line request_id=41908baa-79c1-48e3-acc8-4bd541c9676f team_id=Some(2) distinct_id=Some("0198f39e-af74-7704-9e54-1d7a121a6be8") ip=127.0.0.1 user_agent=Some("posthog-python/3.0.0") lib=Some("posthog-python") lib_version=Some("3.0.0") api_version=Some("2") duration_ms=100 http_status=200 flags_evaluated=Some(6) flags_enabled=Some(5) flags_disabled=false quota_limited=false rate_limited=false error_code=None
2025-12-24T00:20:21.797560Z  INFO ThreadId(05) feature_flags::handler::canonical_log: canonical_log_line request_id=41908baa-79c1-48e3-acc8-4bd541c9676f team_id=Some(2) distinct_id=Some("0198f39e-af74-7704-9e54-1d7a121a6be8") ip=127.0.0.1 user_agent=Some("posthog-python/3.0.0") lib=Some("posthog-python") lib_version=Some("3.0.0") api_version=Some("2") duration_ms=5 http_status=200 flags_evaluated=Some(6) flags_enabled=Some(5) flags_disabled=false quota_limited=false rate_limited=false error_code=None
```

## Changelog: (features only) Is this feature complete?

No - this is internal observability infrastructure, not a user-facing feature.

haacked committed:
Add canonical log line for /flags requests

Implement Stripe-style canonical log lines for the Rust /flags service,
providing a single comprehensive log entry per request for debugging.

Key changes:
- Add CanonicalLogLine struct to accumulate telemetry throughout request
- Add CanonicalLogGuard RAII wrapper to ensure logs emit on early returns
- Add UserAgentInfo module for parsing SDK info from User-Agent headers
- Add error_code() and status_code() methods to FlagError for logging
- Use static str for SDK names to avoid allocations

Fields captured: request_id, team_id, distinct_id, ip, user_agent, lib,
lib_version, api_version, duration_ms, http_status, flags_evaluated,
flags_enabled, flags_disabled, quota_limited, rate_limited, error_code

haacked committed:
Add granular rate limit error types and truncate user-agent in logs

- Add IpRateLimited and TokenRateLimited variants to ClientFacingError
  for more precise error codes in canonical logs
- Update endpoint to use new error variants instead of hardcoded strings
- Truncate user_agent to 512 chars in canonical log to prevent log bloat
  from pathological User-Agent headers

haacked commented on rust/feature-flags/src/api/endpoint.rs:
```diff
@@ -285,14 +295,23 @@ pub async fn flags(
     if !state.flags_rate_limiter.allow_request(&rate_limit_key) {
+        guard.log_mut().rate_limited = true;
+        guard.log_mut().http_status = 429;
+        guard.log_mut().error_code = Some("token_rate_limited".to_string());
```
Fixed.

gustavohstrassburger commented on rust/feature-flags/src/api/errors.rs:
```diff
@@ -108,6 +108,87 @@ pub enum FlagError {
 
 impl FlagError {
+    /// Returns a short error code for canonical logging.
+    pub fn error_code(&self) -> &'static str {
```
We don't need to do this now, but we can probably use canonical logs in other parts of rust code, and some of these could probably be centralized later on.

gustavohstrassburger commented on rust/feature-flags/src/handler/canonical_log.rs:
```diff
@@ -0,0 +1,336 @@
+///
+/// Use [`CanonicalLogGuard`] to ensure the log is always emitted, even on early returns.
+#[derive(Debug, Clone)]
+pub struct CanonicalLogLine {
```
Up to you, not a blocker.

If we decide to expand the use of canonical logs later on, we'll probably have multiple structs, so we could already use a name that references flags, like `FlagsCanonicalLogLine`.

gustavohstrassburger commented on rust/feature-flags/src/utils/user_agent.rs:
```diff
@@ -0,0 +1,378 @@
+        }
+
+        // Check for common HTTP clients (server-side indicators)
+        if ua.contains("python-requests/") || ua.contains("curl/") {
```
Since we don't have support for flags for rust and capacitor, do you think it's valid to to find a way of detecting these languages here in case we have users using flags going around the SDK? 

https://posthog.com/docs/libraries

haacked replied:
Afaik, there's no consistent user-agent for Rust HTTP clients (`reqwest`, `hyper`, `ureq` all differ).

As I understand it, Capacitor uses web views so they'd already be detected as `posthog-js` or `browser`.

gustavohstrassburger reviewed (approved):
Well done, 
I've made some comments, none is a blocker though.

haacked committed:
Fix duplicate canonical log lines on rate limit errors

Calling emit_for_error() through guard.log_mut() emits the log but
doesn't set the guard's emitted flag. When the guard drops on return,
it emits again causing duplicate log lines.

Add set_error() method to set fields without emitting, for use when
the guard will emit on drop. Refactor emit_for_error() to reuse it.

haacked committed:
Rename CanonicalLog* to FlagsCanonicalLog*

Prepares for future expansion of canonical logging to other areas
by using a more specific name for the flags-related structs.

haacked commented on rust/feature-flags/src/handler/canonical_log.rs:
```diff
@@ -0,0 +1,345 @@
+        let duration_ms = self.start_time.elapsed().as_millis() as u64;
+
+        // Truncate distinct_id to prevent log line explosion from long IDs.
+        let distinct_id = self.distinct_id.as_ref().map(|d| truncate_chars(d, 64));
```
Good catch!

It turns out, we already truncate distinct_id in `feature-flags/src/flags/flag_request.rs` lines 170-173, so we don't need to do any truncation here.

```rust
        match distinct_id.len() {
            0..=200 => Ok(distinct_id.to_owned()),
            _ => Ok(distinct_id.chars().take(200).collect()),
        }
```

haacked committed:
Remove redundant distinct_id truncation from canonical log

The distinct_id is already truncated to MAX_DISTINCT_ID_LEN (200 chars)
at request parsing time in flag_request.rs. The additional 64-char
truncation in canonical_log.rs was unnecessarily limiting debug info.

Also extracts the 200-char limit into a named constant.

haacked commented on rust/feature-flags/src/handler/mod.rs:
```diff
@@ -32,16 +34,19 @@ use common_metrics::{histogram, inc};
-pub async fn process_request(context: RequestContext) -> Result<FlagsResponse, FlagError> {
+pub async fn process_request(
+    context: RequestContext,
+    mut canonical_log: FlagsCanonicalLogLine,
```
Hmm, the more I think about it. I like this idea.

haacked committed:
Add experience continuity count to canonical log

Replace flags_enabled with flags_experience_continuity which tracks
how many flags have experience continuity enabled. Also change
flags_evaluated from Option<usize> to usize since 0 is the correct
value when no flags are evaluated.

haacked committed:
Add task_local canonical log with deep evaluation metrics

Migrate canonical log from explicit parameter passing to tokio::task_local,
enabling logging from deep in the flag evaluation stack without threading
parameters through 10+ function levels.

New fields added to canonical log:
- db_property_fetches: database roundtrip count
- property_cache_hits/misses: cache statistics
- cohorts_evaluated: number of cohort filters evaluated
- flags_errored: count of flags with evaluation errors
- hash_key_override_attempted/succeeded: experience continuity tracking

Key changes:
- Add with_canonical_log() helper that no-ops outside scope
- Add run_with_canonical_log() to establish request scope
- Remove FlagsCanonicalLogGuard (replaced by task_local scope)
- Remove canonical_log parameter from process_request()

haacked committed:
Extract rate_limit_error helper to reduce repetition

Consolidates the repeated pattern of updating the canonical log
with rate limit info into a single helper function.

haacked committed:
Clean up canonical log tests and remove unused function

- Remove unused set_canonical_log_error helper function
- Simplify clone test (Clone correctness is compiler-verified)
- Add parameterized truncate_chars tests for edge cases:
  empty strings, boundary conditions, multi-byte emoji, long inputs

haacked committed:
Use Cow for truncate_chars to avoid allocation in common case

Most user-agents are 100-200 chars, well under the 512 limit.
By returning Cow<str>, we avoid ~200 bytes allocation per request
when no truncation is needed.

haacked committed:
Use &'static str for error_code to avoid allocation

FlagError::error_code() returns &'static str, so we can store
it directly without calling .to_string() on every error path.

haacked committed:
Add comprehensive tests and granular DB query metrics

- Add parameterized tests for set_error() and emit_for_error() methods
- Add error code uniqueness test to prevent duplicate codes
- Consolidate error_code()/status_code() into single error_metadata() method
- Add task-local isolation tests for concurrent request safety
- Add granular DB query metrics: person_queries, group_queries, static_cohort_queries

haacked committed:
Add boolean cache miss flags to canonical log

Add person_properties_not_cached and group_properties_not_cached fields
to enable ClickHouse queries like "give me all request_ids where
cache_miss=true and team_id=123" for debugging specific requests.

haacked committed:
Optimize truncate_chars to avoid allocation in all cases

Use char_indices().nth() to find the byte boundary in O(max_chars)
instead of counting all chars first. This also allows returning &str
directly since we can slice the original string at the byte boundary.

Adds more multibyte UTF-8 test cases for emoji and accented characters.

haacked committed:
Combine consecutive with_canonical_log calls into single closures
